src/thing/add.ts

```diff
@@ -56,9 +62,16 @@ export const addUrl: AddOfType<Url | UrlString | Thing> = (
   url
 ) => {
   internal_throwIfNotThing(thing);
+  if (!internal_isValidUrl(property)) {
+    throw new ValidPropertyUrlExpectedError(property);
+  }
   const predicateNode = asNamedNode(property);
   const newThing = internal_cloneThing(thing);
 
+  if (!isThing(url) && !internal_isValidUrl(url)) {
```

Comment:
isn't testing for `isThing(url)` redundant with `  internal_throwIfNotThing(thing);` ?

Reply:
`throwIfNotThing` throws if the Thing that was passed wasn't actually a Thing. This `isThing()` check checks whether the _target_ is a Thing (because you can pass either a Thing or a URL value, and we only need to check the latter for being a valid URL).

In other words:

```
const updatedThing = setUrl(
  thing, // Throws if this is not a Thing
  predicate,
  targetThing, // Adds a link to the given Thing
);
```

and

```

const updatedThing = setUrl(
  thing, // Throws if this is not a Thing
  predicate,
  targetUrl, // Checks that it's not a Thing, and then validates the URL
);
```

Reply:
Aah right of course, thanks